src/python_testing/TC_EWATERHTRM_2_1.py

```diff
@@ -0,0 +1,164 @@
+from mobly import asserts
+
+
+class TC_EWATERHTRM_2_1(MatterBaseTest):
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.endpoint = 0
+
+    def steps_TC_EWATERHTRM_2_1(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "Commissioning, already done", is_commissioning=True),
+            TestStep(2, "Read the SupportedModes attribute"),
+            TestStep(3, "Read the CurrentMode attribute"),
+            TestStep(4, "Send ChangeToMode command with NewMode"),
+            TestStep(5, "Manually put the device in a state from which it will FAIL to transition"),
+            TestStep(6, "Read CurrentMode attribute"),
+            TestStep(7, "Send ChangeToMode command with NewMode"),
+            TestStep(8, "Read CurrentMode attribute"),
+            TestStep(9, "Manually put the device in a state from which it will SUCCESSFULLY transition"),
+            TestStep(10, "Read CurrentMode attribute"),
+            TestStep(11, "Send ChangeToMode command with NewMode"),
+            TestStep(12, "Read CurrentMode attribute"),
+            TestStep(13, "Send ChangeToMode command with NewMode set to an invalid mode"),
+            TestStep(14, "Read CurrentMode attribute"),
+        ]
+        return steps
+
+    async def read_mode_attribute_expect_success(self, endpoint, attribute):
+        cluster = Clusters.Objects.WaterHeaterMode
+        return await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=attribute)
+
+    async def send_change_to_mode_cmd(self, newMode) -> Clusters.Objects.WaterHeaterMode.Commands.ChangeToModeResponse:
+        ret = await self.send_single_cmd(cmd=Clusters.Objects.WaterHeaterMode.Commands.ChangeToMode(newMode=newMode), endpoint=self.endpoint)
+        asserts.assert_true(type_matches(ret, Clusters.Objects.WaterHeaterMode.Commands.ChangeToModeResponse),
+                            "Unexpected return type for Water Heater Mode ChangeToMode")
+        return ret
+
+    def pics_TC_EWATERHTRM_2_1(self) -> list[str]:
+        return ["EWATERHTRM.S"]
+
+    @async_test_body
+    async def test_TC_EWATERHTRM_2_1(self):
+
+        # Valid modes. Only ModeManual referred to in this test
+        # ModeOff    = 0
+        ModeManual = 1
+        # ModeTimed  = 2
+
+        self.endpoint = self.matter_test_config.endpoint
+
+        attributes = Clusters.WaterHeaterMode.Attributes
+
+        self.step(1)
+        # Commission DUT - already done
+
+        self.step(2)
+
+        supported_modes = await self.read_mode_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.SupportedModes)
+
+        logging.info(f"SupportedModes: {supported_modes}")
+
+        asserts.assert_greater_equal(len(supported_modes), 2,
+                                     "SupportedModes must have at least two entries!")
+
+        modes = [m.mode for m in supported_modes]
```

Comment:
please move this closer to the usage - it should be right before the invalid_mode setting.

Reply:
test case seems that it should check that this has 2 or more entries. are those missing?

Reply:
Code modified for first point.
Wrt the second point, we check `supported_modes` has 2 or entries on lines 97/98. Are you looking for a test that checks len(modes) >= 2?

Reply:
Marking as resolved